### PR TITLE
Refactored tests

### LINKED ISSUE #26: Test suite produces warnings
I noticed that after merging #25 there are some tests that produce warnings. By looking at the whole test suite closed I noticed that abstract classes are being used in tests which I think makes the test suite hard to understand. It also causes some tests cases not being executed, more particularly the ones that use `@dataProvider`.

I'd like to propose to refactor the test suite so that abstract classes are not used anymore. This however means that we have duplicate code (copy all test cases into each adapters test class) or use a test method per adapter type and move real assertions to protected methods.

@NoelDavies whats you view on this? I'm happy to submit a example PR if that would help.

### PR DESCRIPTION
**Pull request by martinssipenko**
This PR reorganizes tests so that APCu and BlackBox tests are executed during CI flow.

Closes #26

### CODE REVIEW

**Comment by martinssipenko** [comment]
@NoelDavies I'm also keen on moving the contents of `src/Prometheus/` directory to `src/` as the `Prometheus` is obsolete when using PSR-4. Do  you reckon I should do it in this PR?

**Comment by martinssipenko** [comment]
@NoelDavies any updates on this?

**Comment by NoelDavies** [comment]
> @NoelDavies I'm also keen on moving the contents of `src/Prometheus/` directory to `src/` as the `Prometheus` is obsolete when using PSR-4. Do you reckon I should do it in this PR?

I'd probably create a new PR for this and add it to a major release as people maybe including it directly. or have custom autoloaders, etc.

**Comment by martinssipenko** [comment]
> > @NoelDavies I'm also keen on moving the contents of `src/Prometheus/` directory to `src/` as the `Prometheus` is obsolete when using PSR-4. Do you reckon I should do it in this PR?
> 
> I'd probably create a new PR for this and add it to a major release as people maybe including it directly. or have custom autoloaders, etc.

In that case this PR is ready for review.

File: php-fpm/Dockerfile [hunk 1]
```diff
@@ -1,7 +1,7 @@
-FROM php:5.6-fpm
+FROM php:7.4-fpm
 
-RUN pecl install redis-2.2.8 && docker-php-ext-enable redis
-RUN pecl install apcu-4.0.11 && docker-php-ext-enable apcu
+RUN pecl install redis && docker-php-ext-enable redis
+RUN pecl install apcu && docker-php-ext-enable apcu
```
Comment (NoelDavies) [inline php-fpm/Dockerfile:4]: Shouldn't these be versioned?

File: src/Prometheus/PushGateway.php [hunk 2]
```diff
@@ -71,13 +71,13 @@ public function delete(string $job, array $groupingKey = []): void
...
     /**
-     * @param CollectorRegistry $collectorRegistry
+     * @param null|CollectorRegistry $collectorRegistry
      * @param string $job
      * @param array $groupingKey
      * @param string $method
      * @throws GuzzleException
      */
-    private function doRequest(CollectorRegistry $collectorRegistry, string $job, array $groupingKey, $method): void
+    private function doRequest(?CollectorRegistry $collectorRegistry, string $job, array $groupingKey, $method): void
```
Comment (NoelDavies) [inline src/Prometheus/PushGateway.php:80]: First argument shouldn't be nullable

**Review by NoelDavies — COMMENTED** [review]
I've reviewed a few things, let me know what you think.

File: php-fpm/Dockerfile (hunk 1 above)
Comment (martinssipenko) [reply php-fpm/Dockerfile:4]: I don't see any reason why they should be, I think using latest should be fine for testing, or you think it matters?

File: src/Prometheus/PushGateway.php (hunk 2 above)
Comment (martinssipenko) [reply src/Prometheus/PushGateway.php:80]: When using `delete` method it internally calls `doRequest` with first argument as `null`, check it out [here](https://github.com/martinssipenko/prometheus_client_php/blob/7f4bcc4c95d407e7cc0b8a3e7a29ffb67c4be5eb/src/Prometheus/PushGateway.php#L70) and [here](https://github.com/martinssipenko/prometheus_client_php/blob/7f4bcc4c95d407e7cc0b8a3e7a29ffb67c4be5eb/src/Prometheus/PushGateway.php#L97-L100).

I guess an alternative would be to refactor `delete` to not use `doRequest` but IMHO thats not worth it given `doRequest` is a private method.

**Comment by martinssipenko** [comment]
Any updates on this? Anything I can do to get it merged?

**Comment by NoelDavies** [comment]
@martinssipenko One of the guys at End should be able to help you out. But again, I no longer work there - Apologies.

**Pull request closed without merging**